service/src/java/org/apache/hive/service/cli/operation/ExecuteStatementOperation.java

```diff
@@ -45,6 +62,21 @@ public static ExecuteStatementOperation newExecuteStatementOperation(HiveSession
       throws HiveSQLException {
 
     String cleanStatement = HiveStringUtils.removeComments(statement);
+    if (!HPLSQL.equals(confOverlay.get(QUERY_EXECUTOR)) && hplSqlMode()) {
+      if (SessionState.get().getHplsqlInterpreter() == null) {
+        Exec interpreter = new Exec(
+                new Conf(),
+                new BeelineConsole(),
+                ResultListener.NONE,
+                new HplSqlQueryExecutor(parentSession),
+                parentSession.getMetaStoreClient(),
+                new HiveHplSqlSessionState(SessionState.get())
```

Comment:
The session state is not available when the operation is running in the background.

Reply:
So this gets a reference to active SessionState. Then it is offloaded to another background thread to run. What if the next operation changes sessionState.currentDatabase? Then hpl operation might run on wrong database, no?

Reply:
Yes, that might be a problem, let me check that.

Reply:
@mustafaiman,
I added a unittest to verify it and looks like it still works this way. HiveHplSqlSessionState holds a reference to SessionState which changes when the the current db is modified. The reason why the HiveHplSqlSessionState is still needed is because the SessionState is in the ql project which cannot be referenced from hplsql without introducing circular dependency.